pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +29,66 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, file_obj, relative_url=None):
+        """Upload a file to this repository.
+
+        Args:
+            file_obj (str, file object)
+                If it's a string, then it's the path of file to upload.
+                Else, it ought to be a file-like object
+
+            relative_url (str)
+                Path that should be used in remote repository
+
+                If omitted, filename will be used.
+        Returns:
+            Future[:class:`~pubtools.pulplib.Task`]
+                A future which is resolved when import succeeds.
+
+                The future contains the task to import uploaded content
+                to repository
+
+        Raises:
+            DetachedException
+                If this instance is not attached to a Pulp client.
+        """
+        if not self._client:
+            raise DetachedException()
+
+        relative_url = self._get_relative_url(file_obj, relative_url)
+
+        # request upload id and wait for it
+        upload_id = self._client._request_upload().result()["upload_id"]
+
+        upload_complete_f, checksum, size = self._client._do_upload_file(
+            upload_id, self.id, file_obj
+        )
+
+        unit_key = {"name": relative_url, "digest": checksum, "size": size}
+
+        import_complete_f = f_flat_map(
+            upload_complete_f,
+            lambda _: self._client._do_import(self.id, upload_id, "iso", unit_key),
+        )
+
+        f_map(
+            import_complete_f, lambda _: self._client._delete_upload_request(upload_id)
+        )
+
+        return import_complete_f
+
+    def _get_relative_url(self, file_obj, relative_url):
+        is_path = isinstance(file_obj, str)
+        if is_path:
+            if not relative_url:
+                relative_url = file_obj
+            elif relative_url.endswith("/"):
+                _, name = os.path.split(file_obj)
+                relative_url = os.path.join(relative_url, name)
+        elif not is_path and (not relative_url or relative_url.endswith("/")):
+            msg = "Must provide complete relative_url if the file's not from disk"
+            LOG.exception(msg)
```

Comment:
I don't think LOG.exception is correct here, as I understand it the point of exception() is that it'll dump traceback of the exception currently being handled - which can be also done via any other logger methods by including exc_info=1.

Since you're not yet handling an exception here, it doesn't seem to make sense.  And the message by itself doesn't contain enough info to know which piece of data caused the problem.  Probably just raising here is fine and letting whoever catches the exception worry about logging it.

Also, as a developer I would likely find useful a more detailed and complete message like:

```
"ValueError: %s is missing a `name` attribute and `relative_url` was not provided", file_obj
```

Reply:
A question here, if a file object has the `name` attribute, are we supposed to use that as remote path?